File: base/sources/nodes/string_node.ts

type string_node_t = {
	base?: logic_node_t;
	value?: string;
};

function string_node_create(raw: ui_node_t, args: f32_array_t): string_node_t {
	let n: string_node_t = {};
	n.base = logic_node_create(n);
	n.base.get = string_node_get;
	n.base.set = string_node_set;
	n.value = args == null ? "" : sys_buffer_to_string(args.buffer);
	return n;
}

function string_node_get(self: string_node_t, from: i32): logic_node_value_t {
	if (self.base.inputs.length > 0) {
		return logic_node_input_get(self.base.inputs[0]);
	}
	else {
		let v: logic_node_value_t = { _str: self.value };
		return v;
	}
}

function string_node_set(self: string_node_t, value: any) {
	if (self.base.inputs.length > 0) {
		logic_node_input_set(self.base.inputs[0], value);
	}
	else {
		self.value = value;
	}
}

File: paint/sources/brush_nodes/string_node.c

#include "../global.h"

logic_node_value_t *string_node_get(string_node_t *self, i32 from) {
	if (self->base->inputs->length > 0) {
		return logic_node_input_get(self->base->inputs->buffer[0]);
	}
	else {
		logic_node_value_t *v = GC_ALLOC_INIT(logic_node_value_t, {._str = self->value});
		return v;
	}
}

void string_node_set(string_node_t *self, void * value) {
	if (self->base->inputs->length > 0) {
		logic_node_input_set(self->base->inputs->buffer[0], value);
	}
	else {
		self->value = string_copy(value);
	}
}

string_node_t *string_node_create(ui_node_t *raw, f32_array_t *args) {
	string_node_t *n = GC_ALLOC_INIT(string_node_t, {0});
	n->base          = logic_node_create(n);
	n->base->get     = string_node_get;
	n->base->set     = string_node_set;
	n->value         = args == NULL ? "" : sys_buffer_to_string(args->buffer);
	return n;
}
